raiden/network/transport/matrix/utils.py

```diff
@@ -387,126 +396,134 @@ def join_broadcast_room(client: GMatrixClient, broadcast_room_alias: str) -> Roo
+    # prevents the impersonation metioned above. subsequent calls will reuse
+    # the authentication token and the display name will be properly set.
     signature_bytes = signer.sign(client.user_id.encode())
     signature_hex = encode_hex(signature_bytes)
+
     user = client.get_user(client.user_id)
     user.set_display_name(signature_hex)
+
     log.debug(
-        "Matrix user login", homeserver=server_name, server_url=server_url, username=username
+        "Logged to a new server",
+        node=to_checksum_address(username),
+        homeserver=server_name,
+        server_url=server_url,
     )
     return user
 
 
+def is_valid_username(username: str, server_name: str, user_id: str) -> bool:
+    _match_user = re.match(f"^@{re.escape(username)}:{re.escape(server_name)}$", user_id)
+    return bool(_match_user)
+
+
+def login_with_token(client: GMatrixClient, user_id: str, access_token: str) -> Optional[User]:
+    """Reuse an existing authentication code.
+
+    If this succeeds it means the user has logged in the past, so we assume the
+    display name is properly set and that there may be rooms open from past
+    executions.
+    """
+    client.set_access_token(user_id=user_id, token=access_token)
+
+    try:
+        # Test the credentional. Any API that requries authentication
+        # would be enough.
+        client.api.get_devices()
+    except MatrixRequestError as ex:
+        log.debug(
+            "Couldn't use previous login credentials",
+            node=node_address_from_userid(client.user_id),
+            prev_user_id=user_id,
+            _exception=ex,
+        )
+        return None
+
+    log.debug(
+        "Success. Valid previous credentials",
+        node=node_address_from_userid(client.user_id),
+        user_id=user_id,
+    )
+    return client.get_user(client.user_id)
+
+
+def login(client: GMatrixClient, signer: Signer, prev_auth_data: Optional[str] = None) -> User:
+    """ Login with a matrix server.
+
+    Params:
+        client: GMatrixClient instance configured with desired homeserver.
+        signer: Signer used to sign the password and displayname.
+        prev_auth_data: Previously persisted authentication using the format "{user}/{password}".
+    """
+    server_url = client.api.base_url
+    server_name = urlparse(server_url).netloc
+
+    username = str(to_normalized_address(signer.address))
+
+    if prev_auth_data and prev_auth_data.count("/") == 1:
+        user_id, _, access_token = prev_auth_data.partition("/")
+
+        if is_valid_username(username, server_name, user_id):
+            user = login_with_token(client, user_id, access_token)
+
+            if user is not None:
+                return user
+        else:
+            log.debug(
+                "Auth data is invalid, discarding",
+                node=username,
+                user_id=user_id,
+                server_name=server_name,
+            )
+
+    return first_login(client, signer, username)
```

Comment:
What if the user doesn't exist at all? is the plugin we're using handling this case?

Reply:
https://github.com/raiden-network/raiden/blob/991177cfc5c1c0cc580cf561f52ea6955b3e8cba/raiden/network/transport/matrix/utils.py#L402-L405